cpp/src/arrow/sparse_tensor.cc

```diff
@@ -117,21 +129,36 @@ class SparseTensorConverter<TYPE, SparseCOOIndex>
 
     // make results
     const std::vector<int64_t> indices_shape = {nonzero_count, ndim};
-    const int64_t indices_elsize = sizeof(int64_t);
     const std::vector<int64_t> indices_strides = {indices_elsize,
                                                   indices_elsize * nonzero_count};
-    sparse_index =
-        std::make_shared<SparseCOOIndex>(std::make_shared<SparseCOOIndex::CoordsTensor>(
-            indices_buffer, indices_shape, indices_strides));
+    sparse_index = std::make_shared<SparseCOOIndex>(std::make_shared<Tensor>(
+        index_value_type_, indices_buffer, indices_shape, indices_strides));
     data = values_buffer;
 
     return Status::OK();
   }
 
+#define CALL_TYPE_SPECIFIC_CONVERT(TYPE_CLASS) \
+  case TYPE_CLASS##Type::type_id:              \
+    return Convert<TYPE_CLASS##Type>();
+
+  Status Convert() {
+    switch (index_value_type_->id()) {
+      ARROW_GENERATE_FOR_ALL_INTEGER_TYPES(CALL_TYPE_SPECIFIC_CONVERT);
+      // LCOV_EXCL_START: The following invalid causes program failure.
```

Comment:
Not sure I understand: how does returning `Status::Invalid` cause program failure?

Reply:
Because the result of `Convert` function is checked by `ARROW_CHECK_OK` in `MakeSparseTensorFromTensor` function, and `ARROW_CHECK_OK` makes the fatal-level `ArrowLog` object that makes the program failure.